pkg/view/tui/dependency.go

```diff
@@ -17,84 +17,97 @@
-			Default: false,
-		}, &resp)
-
-		if !resp {
-			return fmt.Errorf("pulumi is required to run this command. For installation instructions see: https://www.pulumi.com/docs/get-started/install/")
-		}
+func DockerBuildxAvailable() (bool, error) {
+	if dockerBuildxAvailable {
+		return true, nil
+	}
 
-		var installErr error
-		platform := runtime.GOOS
-		switch platform {
-		case "darwin", "linux":
-			cmd := exec.Command("sh", "-c", "curl -fsSL https://get.pulumi.com | sh")
-			cmd.Stdout = os.Stdout
-			installErr = cmd.Run()
-		case "windows":
-			cmd := exec.Command("powershell", "-NoProfile", "-InputFormat", "None", "-ExecutionPolicy", "Bypass", "-Command", "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12; iex ((New-Object System.Net.WebClient).DownloadString('https://get.pulumi.com/install.ps1'))")
-			cmd.Stdout = os.Stdout
-			installErr = cmd.Run()
-			if installErr == nil {
-				cmd := exec.Command("powershell", "SET", "PATH='%PATH%;%USERPROFILE%\\.pulumi\\bin'")
-				cmd.Stdout = os.Stdout
-				installErr = cmd.Run()
-			}
-		default:
-			installErr = fmt.Errorf("platform %s not supported", platform)
-		}
+	err := exec.Command("docker", "buildx", "version").Run()
+	if err == nil {
+		dockerBuildxAvailable = true
+	}
+
+	return dockerBuildxAvailable, err
+}
+
+func PodmanAvailable() (bool, error) {
+	if podmanAvailable {
+		return true, nil
+	}
 
-		return installErr
-	},
+	err := exec.Command("podman", "version").Run()
+	if err == nil {
+		podmanAvailable = true
+	}
+
+	return podmanAvailable, err
 }
 
-var Docker = &Dependency{
-	name:    "Docker",
-	command: "docker version",
-	assist: func() error {
-		return fmt.Errorf("docker is required to run this command. For installation instructions see: https://docs.docker.com/engine/install/")
-	},
+type Dependency interface {
+	// Check if the dependency is met
+	Check() error
+	// If the dependency is not met, provide a message to the user to assist in installing it
+	Assist() string
 }
 
-var DockerBuildx = &Dependency{
-	name:    "Docker Buildx",
-	command: "docker buildx version",
-	assist: func() error {
-		return fmt.Errorf("docker buildx is required to run this command. For installation instructions see: https://github.com/docker/buildx")
-	},
+type ContainerToolDependency struct {
+	message string
+}
+
+func (c *ContainerToolDependency) Check() error {
+	_, err := DockerAvailable()
+	if err != nil {
+		_, podErr := PodmanAvailable()
+		if podErr != nil {
+			c.message = "Docker or Podman is required, see https://docs.docker.com/engine/install/ for docker installation instructions"
+			return err
+		} else {
+			// Use podman
+			return nil
+		}
+	}
+
+	_, err = DockerBuildxAvailable()
+	if err != nil {
+		c.message = "docker buildx is required to run this command. For installation instructions see: https://github.com/docker/buildx"
+		return err
+	}
+
+	return nil
+}
+
+func (c *ContainerToolDependency) Assist() string {
+	return c.message
 }
```

Comment:
Could dependency checks be a function with a new error type?

e.g.

```go
type DependencyMissingError struct {
	message string
}

func (d *DependencyMissingError) Error() string {
	return d.message
}

type Dependency = func() error

func ContainerToolDependency() error {
	_, err := DockerAvailable()
	if err != nil {
		_, podErr := PodmanAvailable()
		if podErr != nil {
			return &DependencyMissingError{
				message: "Docker or Podman is required, see https://docs.docker.com/engine/install/ for docker installation instructions",
			}
		} else {
			// Use podman
			return nil
		}
	}

	_, err = DockerBuildxAvailable()
	if err != nil {
		return &DependencyMissingError{
			message: "docker buildx is required to run this command. For installation instructions see: https://github.com/docker/buildx",
		}
	}

	return nil
}
```

Reply:
I've updated it to make dependencies a function. I kept the error (why we think the dependency is missing) separate from the assist message (how we think it can be fixed). However, we're not really using the error right now. Maybe the design should change to just log that error with a debug log level so people can view it as needed, but we can throw it away and remove the second message from the struct.